packages/core/src/stateUtils.ts

```diff
@@ -1508,87 +1485,179 @@ export function resolveTransition<
+          resolveLog(
+            actionObject as LogAction<TContext, TEvent>,
+            context,
+            _event
+          )
+        );
+        break;
+      case actionTypes.pure:
+        resActions.push(
+          ...toArray(
+            (actionObject as PureAction<TContext, TEvent>).get(
+              context,
+              _event.data
+            )
+          )
+        );
+        break;
+      case actionTypes.assign:
+        context = updateContext(
+          context,
+          _event,
+          [actionObject as AssignAction<TContext, TEvent>],
+          currentState
+        );
+        resActions.push(actionObject);
+        break;
+      default:
+        resActions.push(
+          toActionObject(actionObject, machine.machine.options.actions)
+        );
+        break;
     }
+  });
 
-    while (raisedEvents.length) {
-      const raisedEvent = raisedEvents.shift()!;
-      maybeNextState = resolveRaisedTransition(
-        machine,
-        maybeNextState,
-        raisedEvent._event,
-        _event
-      );
-    }
-  }
+  return {
+    actions: resActions,
+    raised: raisedActions,
+    context
+  };
+}
+
+export function macrostep<TContext, TEvent extends EventObject>(
+  state: State<TContext, TEvent, any, Typestate<TContext>>,
+  event: Event<TEvent> | SCXML.Event<TEvent> | null,
+  machine: MachineNode<TContext, any, TEvent, any>
+): State<TContext, TEvent> {
+  // Assume the state is at rest (no raised events)
+  // Determine the next state based on the next microstep
+  const nextState = event === null ? state : machine.microstep(state, event);
+
+  const { _event, _internalQueue } = nextState;
+  let maybeNextState = nextState;
 
-  // Detect if state changed
-  const changed =
-    maybeNextState.changed ||
-    (history
-      ? !!maybeNextState.actions.length ||
-        !!assignActions.length ||
-        typeof history.value !== typeof maybeNextState.value ||
-        !stateValuesEqual(maybeNextState.value, history.value)
-      : undefined);
+  while (_internalQueue.length && !maybeNextState.done) {
+    const raisedEvent = _internalQueue.shift()!;
+    const currentActions = maybeNextState.actions;
 
-  maybeNextState.changed = changed;
+    maybeNextState = machine.microstep(
+      maybeNextState,
+      raisedEvent as SCXML.Event<TEvent>
+    );
 
-  // TODO: remove children if they are stopped
-  maybeNextState.children = children;
+    _internalQueue.unshift(...maybeNextState._internalQueue);
 
-  // Preserve original history after raised events
-  maybeNextState.history = history;
-  maybeNextState.historyValue = resolved.historyValue;
+    // Save original event to state
+    if (raisedEvent.type === NULL_EVENT) {
+      maybeNextState._event = _event;
+      maybeNextState.event = _event.data;
+    }
+
+    // Since macrostep actions have not been executed yet,
+    // prioritize them in the action queue
+    maybeNextState.actions.unshift(...currentActions);
```

Comment:
i've noticed now that we have a quite serious bug regarding external events order, it's possible to schedule an external event from an arbitrary action or other services **in between** of returned `send` actions - probably a topic for another PR and a round of changes but it's worth noting that this might happen right now

Reply:
After another thought... this is not really a bug. It’s always possible that in response to an action an external event gets scheduled so it’s possible for it to interleave returned send actions. It probably would be harded for it to happen in multithreaded env but this is JS and this situation is fine